### PR TITLE
Support FastSAM directory inference and plot

### PR DESCRIPTION
**Pull request by HtwoOtwo**
### Summary
🖼️🛠️🔄

This pull request adds multi-image support and `Result` integration to the `FastSAMPrompt` class, and refactors the code for better quality and usability. It also modifies some output formats in `ultralytics/models/fastsam/prompt.py`.

> _`FastSAMPrompt` unleashes the power of the mask_
> _Revealing the shapes of doom in the image task_
> _`Result` objects hold the secrets of the past_
> _The code is refined, the output is sublime_

### Walkthrough
*  Modify `FastSAMPrompt` class to accept single image or directory of images as source ([link](https://github.com/ultralytics/ultralytics/pull/4634/files?diff=unified&w=0#diff-525486f27ba5eca08d7cfaca5bdb538fb7c4c7963005d75565cfd5912be42438L11-R20), [link](https://github.com/ultralytics/ultralytics/pull/4634/files?diff=unified&w=0#diff-525486f27ba5eca08d7cfaca5bdb538fb7c4c7963005d75565cfd5912be42438L89-R163), [link](https://github.com/ultralytics/ultralytics/pull/4634/files?diff=unified&w=0#diff-525486f27ba5eca08d7cfaca5bdb538fb7c4c7963005d75565cfd5912be42438L218-R232), [link](https://github.com/ultralytics/ultralytics/pull/4634/files?diff=unified&w=0#diff-525486f27ba5eca08d7cfaca5bdb538fb7c4c7963005d75565cfd5912be42438L240-R257), [link](https://github.com/ultralytics/ultralytics/pull/4634/files?diff=unified&w=0#diff-525486f27ba5eca08d7cfaca5bdb538fb7c4c7963005d75565cfd5912be42438L268-R288))
*  Refactor `_format_results` method to use `continue` statement and improve readability ([link](https://github.com/ultralytics/ultralytics/pull/4634/files?diff=unified&w=0#diff-525486f27ba5eca08d7cfaca5bdb538fb7c4c7963005d75565cfd5912be42438L54-R61))
*  Update `_get_bbox_from_mask` method to calculate bounding box height and width from coordinates ([link](https://github.com/ultralytics/ultralytics/pull/4634/files?diff=unified&w=0#diff-525486f27ba5eca08d7cfaca5bdb538fb7c4c7963005d75565cfd5912be42438R78-R79))
*  Rewrite `plot` method to handle multiple images, use `cv2.imwrite` function, and show progress bar ([link](https://github.com/ultralytics/ultralytics/pull/4634/files?diff=unified&w=0#diff-525486f27ba5eca08d7cfaca5bdb538fb7c4c7963005d75565cfd5912be42438L89-R163))
*  Raise `ValueError` in `_crop_image`, `box_prompt`, and `point_prompt` methods if source is a directory ([link](https://github.com/ultralytics/ultralytics/pull/4634/files?diff=unified&w=0#diff-525486f27ba5eca08d7cfaca5bdb538fb7c4c7963005d75565cfd5912be42438L218-R232), [link](https://github.com/ultralytics/ultralytics/pull/4634/files?diff=unified&w=0#diff-525486f27ba5eca08d7cfaca5bdb538fb7c4c7963005d75565cfd5912be42438L240-R257), [link](https://github.com/ultralytics/ultralytics/pull/4634/files?diff=unified&w=0#diff-525486f27ba5eca08d7cfaca5bdb538fb7c4c7963005d75565cfd5912be42438L268-R288))
*  Update `box_prompt`, `point_prompt`, `text_prompt`, and `everything_prompt` methods to return `results` object instead of numpy array ([link](https://github.com/ultralytics/ultralytics/pull/4634/files?diff=unified&w=0#diff-525486f27ba5eca08d7cfaca5bdb538fb7c4c7963005d75565cfd5912be42438L240-R257), [link](https://github.com/ultralytics/ultralytics/pull/4634/files?diff=unified&w=0#diff-525486f27ba5eca08d7cfaca5bdb538fb7c4c7963005d75565cfd5912be42438L268-R288), [link](https://github.com/ultralytics/ultralytics/pull/4634/files?diff=unified&w=0#diff-525486f27ba5eca08d7cfaca5bdb538fb7c4c7963005d75565cfd5912be42438L288-R303), [link](https://github.com/ultralytics/ultralytics/pull/4634/files?diff=unified&w=0#diff-525486f27ba5eca08d7cfaca5bdb538fb7c4c7963005d75565cfd5912be42438L298-R316))
*  Rename `with_countouers` argument to `withContours` for consistency ([link](https://github.com/ultralytics/ultralytics/pull/4634/files?diff=unified&w=0#diff-525486f27ba5eca08d7cfaca5bdb538fb7c4c7963005d75565cfd5912be42438L89-R163))
*  Remove unused import of `pathlib.Path` ([link](https://github.com/ultralytics/ultralytics/pull/4634/files?diff=unified&w=0#diff-525486f27ba5eca08d7cfaca5bdb538fb7c4c7963005d75565cfd5912be42438L4))

now you can use:
```
# Define an inference source
source = "path to your image folder"

# Create a FastSAM model
model = FastSAM("FastSAM-s.pt")  # or FastSAM-x.pt

# Run inference on an image
everything_results = model(source, device='cpu', retina_masks=True, imgsz=1024, conf=0.4, iou=0.9)
```
for directory inference

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Improve FastSAMPrompt usability and fix potential bugs.

### 📊 Key Changes
- Changed `full_box` tensor initialization to match expected dimensions.
- Replaced `LOGGER` with `TQDM_BAR_FORMAT` for better progress visualization.
- Refactored the code to better handle image sources and results plotting.
- Conditional checks for masks presence to prevent errors.
- Streamlined functions to work correctly with directories as sources.
- Added tqdm progress bars for visual feedback during processing.
- Enhanced image contour drawing and handling within the plotting function.
- Included checks to prevent directory usage where file paths are expected.

### 🎯 Purpose & Impact
- 🛠️ Fixes potential dimension mismatch issues for a smoother run.
- 🔄 Updated the CLI experience with progress indicators, making it easier for users to follow the process.
- 🛡️ Error handling additions prevent crashes when unexpected input is given, improving reliability.
- 📈 Enhanced plotting capabilities offer better quality visual output, leading to improved user satisfaction.
- 🏗️ Code refactoring leads to more maintainable code, facilitating future updates and contributions.

### CODE REVIEW

**Comment by HtwoOtwo** [comment]
> @HtwoOtwo thanks, this looks good!
> 
> Can you replace the os.path usaged with the newer pathlib.Path usage? Also have you tested the plotting and made sure the changes work correctly?

Of course! In addition, the previous `fig.savefig(save_path)` does not support the bmp format, so I changed it to `cv2.imwrite`. The plotting works fine, but it should be noted that plt is not thread-safe so I haven't used multi-threading for the time being. Using prompt when you're doing directory inference is ambiguous, so I added exception handling to disable using prompt when source is a directory. last but not least, I fixed the bug that `full_box = [torch.zeros_like(p[0][0])` will report an error when the post-processing uses nms and the result is empty.

**Comment by glenn-jocher** [comment]
Thanks @HtwoOtwo for the detailed analysis and proposed improvements for the code in YOLOv8. 

Regarding replacement of `os.path` with `pathlib.Path`, I agree with you, the use of `pathlib` provides an object-oriented way to handle filesystem paths which is more efficient and less prone to errors. 

It's great to see that you have done the necessary testing on your changes, especially on the plotting feature. Your point about Matplotlib not being thread-safe is valid and we'll look into possible solutions for it.

The exception handling for disabling prompt when source is a directory seems like a good addition. This should make the user experience much smoother.

The fixes you have implemented for the post-processing bugs will certainly enhance the performance of the system. Thank you for pointing them out.

Your contributions to improve YOLOv8 are appreciated. Let's keep the collaborative effort going to make the system even better.

**Comment by glenn-jocher** [comment]
@HtwoOtwo I think this PR may have introduced a bug as tests are failing for me in certain environments, so I've fixed this in https://github.com/ultralytics/ultralytics/pull/4634

The failing test was `tests/test_cli.py::test_fastsam` on macOS.

**Comment by HtwoOtwo** [comment]
> @HtwoOtwo I think this PR may have introduced a bug as tests are failing for me in certain environments, so I've fixed this in #4705
> 
> The failing test was `tests/test_cli.py::test_fastsam` on macOS on M1/M2 processors.

Thanks for pointing out and fixing this bug!

**Comment by glenn-jocher** [comment]
@HtwoOtwo thanks for catching this and for the quick fix! It's definitely important to ensure that our code is running smoothly across different environments. I'll take note and make sure to be more mindful of this in future contributions. Appreciate your diligence!
